solc/CommandLineInterface.cpp

```diff
@@ -639,7 +639,7 @@ bool CommandLineInterface::run(int _argc, char const* const* _argv)
 		// There might be no message in the exception itself if the error output is bulky and has
 		// already been printed to stderr (this happens e.g. for compiler errors).
 		if (_exception.what() != ""s)
-			serr() << _exception.what() << std::endl;
+			report(Error::Severity::Error, _exception.what());
```

Comment:
Looks like this breaks `--no-color` for errors than happen in `CommandLineParser`. If an exception is thrown there, we won't know if `--no-color` was used. And if it was, we'll ignore it and use autodetection, which in many cases will result in color being used.

Reply:
It seems that some people are using it: https://github.com/foundry-rs/foundry/issues/5254#issuecomment-1615826179.

Reply:
In what cases *do* we get an exception from ``CommandLineParser`` here? Is that only weird stuff or only regular issues like incorrect options?

Reply:
I'd myself rather tend to err on the side of caution here and not colorize output until we parsed options... we can just split into two separate ``try`` blocks for that, can't we? Any reason not to do that? Is it possible to properly account for ``solc --no-color <someErroneaousArgs>``?

Reply:
> In what cases do we get an exception from `CommandLineParser` here? Is that only weird stuff or only regular issues like incorrect options?

Regular stuff. It's any validation performed by `boost::program_options` (e.g. rejecting invalid options) and our general validation of input (conflicting options, options allowed in each input mode, validation of remappings, etc.)

> we can just split into two separate try blocks for that, can't we? Any reason not to do that?

The exception that causes this happens already in `parseArguments()` and when it happens we discard the partially filled struct so splitting won't help. But I solved it by adding another `try` block inside `parseArguments()` to preserve just the color status.

> I'd myself rather tend to err on the side of caution here and not colorize output until we parsed options...

OK. In that case I added a few commits to steer the PR in that direction. Now the `--color`/`--no-color` flags are parsed first and the resulting color status is copied from the partially parsed options struct even if there's an exception. I also made it conservatively disable color if it's not certain if these options were used or in the message displayed when they're both used. See the last 3 commits.